osquery/filesystem/linux/proc.h

```diff
@@ -0,0 +1,281 @@
+Status procProcesses(bool (*callback)(const std::string&, UserData),
+                     UserData data) {
+  boost::filesystem::directory_iterator it(kLinuxProcPath), end;
+
+  try {
+    for (; it != end; ++it) {
+      if (!boost::filesystem::is_directory(it->status())) {
+        continue;
+      }
+
+      // See #792: std::regex is incomplete until GCC 4.9
+      const auto& pid = it->path().leaf().string();
+      if (std::atoll(pid.data()) <= 0) {
+        continue;
+      }
+
+      if (!callback(pid, data)) {
+        break;
+      }
+    }
+
+  } catch (const boost::filesystem::filesystem_error& e) {
+    VLOG(1) << "Exception iterating Linux processes " << e.what();
+    return Status(1, e.what());
+  }
+
+  return Status(0, "OK");
+}
+
+template <typename UserData>
+Status procDescriptors(const std::string& process_id,
+                       bool (*callback)(const std::string&,
+                                        const std::string&,
+                                        UserData),
+                       UserData data) {
+  auto descriptors_path = kLinuxProcPath + "/" + process_id + "/fd";
+
+  try {
+    boost::filesystem::directory_iterator it(descriptors_path), end;
+
+    for (; it != end; ++it) {
+      auto fd = it->path().leaf().string();
+
+      std::string linkname;
+      if (!procReadDescriptor(process_id, fd, linkname).ok()) {
+        continue;
+      }
+
+      if (!callback(fd, linkname, data)) {
+        break;
+      }
+    }
+
+    return Status(0, "OK");
+
+  } catch (boost::filesystem::filesystem_error& e) {
+    return Status(1,
+                  std::string("Cannot access descriptors for ") + process_id);
+  }
+}
+
+template <typename UserData>
+Status procProcessSockets(bool (*callback)(const ProcessSocket&, UserData),
+                          UserData user_data,
+                          const std::string& process_id,
+                          int protocol,
+                          int family,
+                          std::unordered_map<std::string, std::string>*
+                              inode_to_fd_map_ptr = nullptr) {
+  if (protocol != IPPROTO_IP &&
+      kLinuxProtocolNames.find(protocol) == kLinuxProtocolNames.end()) {
+    return Status(1, "Invalid protocol specified");
+  }
+
+  if (family != AF_UNIX && family != AF_INET && family != AF_INET6) {
+    return Status(1, "Invalid address family specified");
+  }
+
+  std::unordered_map<std::string, std::string> inode_to_fd_map;
+  if (inode_to_fd_map_ptr != nullptr) {
+    inode_to_fd_map = *inode_to_fd_map_ptr;
+
+  } else {
+    auto status = procSocketInodeToFdMap(process_id, inode_to_fd_map);
+    if (!status.ok()) {
+      return Status(1, "Failed to enumerate the process descriptors");
+    }
+  }
+
+  auto socket_list_path = kLinuxProcPath + "/" + process_id + "/net/";
+
+  if (family == AF_UNIX) {
+    socket_list_path += "unix";
+  } else {
+    socket_list_path += kLinuxProtocolNames.at(protocol);
```

Comment:
We can only receive AF_UNIX, AF_INET and AF_INET6 (due to the second **if** statement in the function)

```
  if (family != AF_UNIX && family != AF_INET && family != AF_INET6) {
    return Status(1, "Invalid address family specified");
  }
```

Can you help me identify the fault?

Reply:
Maybe I'm getting this wrong but my idea is: if I call this function with `protocol = IPPROTO_IP` and `family = AF_INET` it will get to the `else` block here and `kLinuxProtocolNames.at()` will be called with `protocol = IPPROTO_IP`. `IPPROTO_IP` is not in the `kLinuxProtocolNames` map so `at()` will fail.

In the codebase this will never happen because `IPPROTO_IP` is only used in conjunction with `AF_UNIX` but still we should check.

Reply:
I've added an `if` to make sure that IPPROTO_IP can only be used with the AF_UNIX family